src/ceph-volume/ceph_volume/util/metadata.py

```diff
@@ -0,0 +1,438 @@
+"""
+This module is responsible for building a metadata cache lookup for every devices on the system, and
+may be used as the --with-preload option within the ceph-volume 'inventory' sub-command. The metadata
+is gathered using a producer/consume thread model to try and perform as much processing as possible in
+parallel to reduce runtime. In addition, the LVM queries are run once, returning all data for all
+related objects in place of the LVM calls issued per device within the Device class.
+
+The parallelism via the thread model settled on producer/consumer + queue model for speed (~0.7s). The
+inbuilt concurrent.futures module was tested, but ran at nearly double the latency (~1.3s).
+"""
+
+
+import os
+import glob
+import json
+import time
+import queue
+import logging
+import threading
+
+from typing import List, Dict, Any, Optional
+
+from ceph_volume.process import call
+from ceph_volume.util.lsmdisk import LSMDisk
+from ceph_volume.util.disk import udevadm_property, get_file_contents, blkid
+from ceph_volume.api.lvm import Volume, PVolume, VolumeGroup, LV_FIELDS, PV_FIELDS, VG_FIELDS
+from ceph_volume.util.constants import DEVID_PROPS
+
+logger = logging.getLogger(__name__)
+
+task_queue = queue.Queue()
+shared_data_lock = threading.Lock()
+
+
+class SharedData:
+    def __init__(self):
+        self.udev = {}
+        self.lsm = {}
+        self.blkid = {}
+        self.lv = {}
+        self.pv = {}
+        self.vg = {}
+
+
+shared_data = SharedData()
+
+
+def get_lsm(dev_path: str) -> Dict[str, Any]:
```

Comment:
do we run mypy against ceph-volume?

Reply:
I'd adopted that approach to better document the code

Reply:
would be great to have some mypy coverage!